sql/src/main/java/org/apache/druid/sql/calcite/planner/SqlParameterizerShuttle.java

```diff
@@ -50,34 +63,36 @@ public SqlParameterizerShuttle(PlannerContext plannerContext)
   public SqlNode visit(SqlDynamicParam param)
   {
     try {
-      if (plannerContext.getParameters().size() > param.getIndex()) {
-        TypedValue paramBinding = plannerContext.getParameters().get(param.getIndex());
-        if (paramBinding == null) {
-          throw new IAE("Parameter at position[%s] is not bound", param.getIndex());
-        }
-        if (paramBinding.value == null) {
-          return SqlLiteral.createNull(param.getParserPosition());
-        }
-        SqlTypeName typeName = SqlTypeName.getNameForJdbcType(paramBinding.type.typeId);
-        if (SqlTypeName.APPROX_TYPES.contains(typeName)) {
-          return SqlLiteral.createApproxNumeric(paramBinding.value.toString(), param.getParserPosition());
-        }
-        if (SqlTypeName.TIMESTAMP.equals(typeName) && paramBinding.value instanceof Long) {
-          return SqlLiteral.createTimestamp(
-              TimestampString.fromMillisSinceEpoch((Long) paramBinding.value),
-              0,
-              param.getParserPosition()
-          );
-        }
-
-        return typeName.createLiteral(paramBinding.value, param.getParserPosition());
-      } else {
-        throw new IAE("Parameter at position[%s] is not bound", param.getIndex());
+      if (plannerContext.getParameters().size() <= param.getIndex()) {
+        throw new IAE("Parameter at position [%s] is not bound", param.getIndex());
+      }
+      TypedValue paramBinding = plannerContext.getParameters().get(param.getIndex());
+      if (paramBinding == null) {
+        throw new IAE("Parameter at position [%s] is not bound", param.getIndex());
+      }
+      if (paramBinding.value == null) {
+        return SqlLiteral.createNull(param.getParserPosition());
       }
+      SqlTypeName typeName = SqlTypeName.getNameForJdbcType(paramBinding.type.typeId);
+      if (SqlTypeName.APPROX_TYPES.contains(typeName)) {
+        return SqlLiteral.createApproxNumeric(paramBinding.value.toString(), param.getParserPosition());
+      }
+      if (SqlTypeName.TIMESTAMP.equals(typeName) && paramBinding.value instanceof Long) {
+        return SqlLiteral.createTimestamp(
+            TimestampString.fromMillisSinceEpoch((Long) paramBinding.value),
+            0,
+            param.getParserPosition()
+        );
+      }
+
+      // This throws ClassCastException for a DATE parameter given as
```

Comment:
If this is the only place that's expected to throw a ClassCastException, how about wrapping only this particular line in the try/catch?